src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs

```diff
@@ -406,42 +406,36 @@ internal override bool ApplyNullableTransforms(byte defaultTransformFlag, Immuta
 
         internal override TypeSymbol SetObliviousNullabilityForReferenceTypes()
         {
-            if (ElementType.TypeKind == TypeKind.Array)
-            {
-                // The language supports deeply nested arrays, up to 10,000+ instances. This means we can't implemented a head
-                // recursive solution here. Need to take an iterative approach to building up the annotations here.
-                var builder = ArrayBuilder<ArrayTypeSymbol>.GetInstance();
+            // The language supports deeply nested arrays, up to 10,000+ instances. This means we can't implemented a head
+            // recursive solution here. Need to take an iterative approach to building up the annotations here.
+            var builder = ArrayBuilder<ArrayTypeSymbol>.GetInstance();
 
-                // Dig through the element types to get the set of ArrayTypeSymbol instances that need to be marked
-                // as oblivious
-                var array = this;
-                TypeWithAnnotations elementType;
-                do
-                {
-                    builder.Add(array);
-                    elementType = array.ElementTypeWithAnnotations;
-                    array = elementType.TypeKind == TypeKind.Array
-                        ? (ArrayTypeSymbol)elementType.Type
-                        : null;
-                }
-                while (!(array is null));
-
-                // Fixup the element type on the most nested array
-                elementType = elementType.SetObliviousNullabilityForReferenceTypes();
-                array = (builder[builder.Count - 1]).WithElementType(elementType);
+            // Dig through the element types to get the set of ArrayTypeSymbol instances that need to be marked
+            // as oblivious
+            var array = this;
+            TypeWithAnnotations elementType;
+            do
+            {
+                builder.Push(array);
+                elementType = array.ElementTypeWithAnnotations;
+                array = elementType.Type as ArrayTypeSymbol;
+            }
+            while (!(array is null));
 
-                // All but the most nested array instance is just an oblivious array with the same custom modifiers
-                for (var i = builder.Count - 2; i >= 0; i--)
-                {
-                    var previousArray = builder[i];
-                    elementType = TypeWithAnnotations.Create(array, NullableAnnotation.Oblivious, previousArray.ElementTypeWithAnnotations.CustomModifiers);
-                    array = (ArrayTypeSymbol)(previousArray.WithElementType(elementType));
-                }
+            // Fixup the element type on the most nested array
+            elementType = elementType.SetObliviousNullabilityForReferenceTypes();
+            array = builder.Pop().WithElementType(elementType);
```

Comment:
Is this `Pop()` case necessary of would it be handled by the `while` loop below?

Reply:
Yes it's necessary. The first `WithElementType` is a special case because it needs to be the result of `TypeWithAnnotations.SetObliviousNullabilityFoReferenceTypes`. This specific type is opaque here, or not array, we don't special case it in any way. The rest of the `WithElementType` calls are operating on `ArrayTypeSymbol` instances and they have to manually re-create the `TypeWithAnnotations`

Reply:
Can the fixup be written as:
```C#
elementType = elementType.SetObliviousNullabilityForReferenceTypes();
while (builder.Count > 0)
{
    array = builder.Pop().WithElementType(elementType);
    elementType = TypeWithAnnotations.Create(array, NullableAnnotation.Oblivious, previousArray.ElementTypeWithAnnotations.CustomModifiers);
}
```

---
In reply to: [274030114](https://github.com/dotnet/roslyn/pull/34891#discussion_r274030114) [](ancestors = 274030114)

Reply:
Pretty sure that messes up the custom modifiers. Let's say that arrayA has element type arrayB. In this sample the first `Pop` returns arrayB and sets `elementType` to have the custom modifiers of the original `elementType` (outside the loop). What it really needs though is to take the custom modifiers off of the element on arrayA. 


---
In reply to: [274056977](https://github.com/dotnet/roslyn/pull/34891#discussion_r274056977) [](ancestors = 274056977,274030114)